File: Algorithm/1-TwoPointers/28_strStr.cpp
//12-29
#include <iostream>
#include <string>
using namespace std;

class Solution {
public:
    int strStr(string haystack, string needle) {
        if (needle.size() == 0)
            return 0;
        for (int i = 0; i < haystack.size(); i++) {
            if (haystack.size() - i < needle.size())
                break;
            if (haystack[i] != needle[0])
                continue;
            
            for (int j = i, k = 0; k < needle.size(); j++, k++) {
                if (haystack[j] != needle[k])
                    break;
                if (k == needle.size() - 1)
                    return j - k;
            }
        }
        return -1;
    }
};
// 就是一个一个比 也没什么

File: Algorithm/1-TwoPointers/28_strStr.java
package LeetCode.Algorithm.TwoPointers;

class Solution {
    public int strStr(String haystack, String needle) {
        int L = needle.length(), n = haystack.length();
        for (int start = 0; start < n - L + 1; ++start) {
            if (haystack.substring(start, start + L).equals(needle))
                return start;
        }
        return -1;
    }
}

// 双指针
class Solution {
    public int strStr(String haystack, String needle) {
        int L = needle.length(), n = haystack.size();
        if (L == 0)
            return 0;
        
        int pn = 0;
        while (pn < n - L + 1) {
            while (pn < n - L + 1 && haystack.charAt(pn) != needle.charAt(0))
                pn++;

            int currLen = 0, pL = 0;
            while (pL < L && pn < n && haystack.charAt(pn) == needle.charAt(pL)) {
                ++pn;
                ++pL;
                ++currLen;
            }
            if (currLen == L)
                return pn - L;
            pn = pn - currLen + 1;
        }
        return -1;
    }
}
